File: Python-DSA/bubble_sort.py
# Bubble sort
# Logic: It compares two adjecent element in every pass

# Best case: O(n) { if the input is in asscending order}
# Worst case: O(n^2) { if the input is in descending order}

def bubble_sort(arr):

    flag = False
    
    for i in range(len(arr)-1):
        print("Pass: ", i+1)
        for j in range(len(arr)-i-1):
            if(arr[j]>arr[j+1]):
                arr[j], arr[j+1] = arr[j+1], arr[j]

                flag = True

                print(arr, "Swapped", arr[j], arr[j+1]) # swapped iteration
            else:
                print(arr, "Unswapped", arr[j], arr[j+1]) # unswapped iteration  
        print()
        if (flag == False):
            return arr;
    print('Reached')
    return arr


def main():
    # arr = [34,23,2,44,1,0,9]
    arr = [1,2,3,4,5,6,0]
    print('Unsorted Array: ', arr)

    result = bubble_sort(arr)
    print('Sorted Array: ', result)


if __name__ == "__main__" :
    main()

File: Java-DSA/BubbleSort.java
import java.util.Arrays;

class BubbleSort {
  public static void main(String[] args) {
    System.out.println("Program for bubble sort\n");

    int[] arr = {22,43,21,3,40,1};
    System.out.println("Entered array : "+ Arrays.toString(arr));

    int[] result = bubbleSort(arr);
    System.out.println("Sorted Array: "+ Arrays.toString(result));
  }


  public static int[] bubbleSort(int[]arr) {

    int size = arr.length;
    for(int i = 0; i < size; i++){

      for(int j = 0; j < size-1; j++){
        int temp = 0;
        if(arr[j] > arr[j+1]){
          temp = arr[j];
          arr[j] = arr[j+1];
          arr[j+1] = temp;

        }
      }
    }
    return arr;
  }
}